server/src/main/java/io/pravega/schemaregistry/storage/SchemaStore.java

```diff
@@ -235,21 +237,28 @@
      * @param namespace namespace 
      * @param group group 
      * @param schemaInfo schema to add
+     * @param fingerprint 256 bit sha hash of schema binary. 
+     *                    Two schema binary representation will be considered identical if their fingerprints match.
+     * @param equality checks if a schema info object is equal to the supplied schemainfo. 
      * @param prop group properties applied at the time of schema addition.
      * @param etag entity tag for the group. 
      * @return Completablefuture that holds version info for the schema that is added.  
      */
-    CompletableFuture<VersionInfo> addSchema(String namespace, String group, SchemaInfo schemaInfo, GroupProperties prop, Etag etag);
+    CompletableFuture<VersionInfo> addSchema(String namespace, String group, SchemaInfo schemaInfo, BigInteger fingerprint,
+                                             Predicate<SchemaInfo> equality, GroupProperties prop, Etag etag);
```

Comment:
The parameters of this call make sense, except this equality addition. A general predicate to test for equality in this interface fells extraneous. The question is redundant, but I'm wondering whether we can bring the validation and comparison logic entirely to the implementation of `addSchema` and other methods here rather than let it leak through the interface.

Reply:
I understand the intent, but it will break the abstraction. 
The schema is opaque bytes to the store layer. its job is to simply store the schema. 
If we assume that a sha-256 hash is sufficient and two schemas with same hash will be considered identical (which is true for all practical purposes as the probability of a collision is negligible), then we dont need to pass such a predicate (or the normalized schema form as we were doing in the previous iteration).